### PR TITLE
CB-15368 Enable FreeIPA HA upgrade

### PR DESCRIPTION
**Pull request by lacikaaa**
- modified validation to allow upgrade up to 3 instances
- refactored `FreeIpaTopologyService`

Tested upgrade for FreeIPA with 3 instances, and validated topology is fine when 4 instances are present. Also started DL after upgrade.

See detailed description in the commit message.

### CODE REVIEW

**Comment by lacikaaa** [comment]
@jamisonbennett I had refactored the `FreeIpaTopologyService` iterator based solution as it was a bit hard to follow for me. I haven't touched the tests, and the refactored version produces the same output.

File: freeipa/src/test/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaTopologyServiceTest.java [hunk 1]
```diff
@@ -36,13 +38,10 @@
...
-
-    @Mock
-    private Stack stack;
+    private InstanceMetaDataService instanceMetaDataService;
 
     @Test
-    void testGenerateTopology() throws FreeIpaClientException {
+    void testGenerateTopology() {
         Set<String> nodes = new HashSet<>();
         assertSetEquivlance(underTest.generateTopology(Set.of("")), Set.of());
```
Comment (gergopapi2) [inline freeipa/src/test/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaTopologyServiceTest.java:46]: Typo: Equivalence

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaTopologyService.java [hunk 2]
```diff
@@ -34,69 +36,75 @@
...
-                    s.setRightNode(pair.getRight());
-                    s.setCn(String.format("%s-to-%s", pair.getLeft(), pair.getRight()));
-                    return s;
-                }).collect(Collectors.toSet());
+                .map(this::convertUnorderedPairToTopologySegment)
+                .collect(Collectors.toSet());
         List<TopologySuffix> topologySuffixes = freeIpaClient.findAllTopologySuffixes();
         for (TopologySuffix topologySuffix : topologySuffixes) {
             createMissingSegments(freeIpaClient, topologySuffix.getCn(), topology);
             removeExtraSegements(freeIpaClient, topologySuffix.getCn(), topology);
```
Comment (gergopapi2) [inline freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaTopologyService.java:53]: Typo: Segments

**Review by gergopapi2 — CHANGES_REQUESTED** [review]

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaTopologyService.java [hunk 3]
```diff
@@ -34,69 +36,75 @@
     private static final Integer MAX_REPLICATION_CONNECTIONS = 4;
 
     @Inject
-    private StackService stackService;
+    private InstanceMetaDataService instanceMetaDataService;
```
[the remark this answers is not in the document]
Comment (lacikaaa) [reply freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaTopologyService.java:39]: I totally agree with it, but I didn't want to do it in one step, so the unit test would ensure that my refactor produces the same output. Once it's merged we can split this class into multiple one to clean it up a bit more, and in this case a javadoc might be helpful, but I think it's a great thing that the unit test helped to explain the logic :)

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/upgrade/UpgradeValidationService.java [hunk 4]
```diff
@@ -22,16 +22,18 @@
...
+
     @Inject
     private EntitlementService entitlementService;
 
     public void validateStackForUpgrade(Set<InstanceMetaData> allInstances, Stack stack) {
         if (allInstances.isEmpty()) {
             LOGGER.warn("Instances are empty for stack.");
             throw new BadRequestException("There are no instances available for upgrade!");
-        } else if (allInstances.size() > 2) {
+        } else if (allInstances.size() > MAX_NUMBER_OF_INSTANCES_FOR_UPGRADE) {
```
[the remark this answers is not in the document]
Comment (lacikaaa) [reply freeipa/src/main/java/com/sequenceiq/freeipa/service/upgrade/UpgradeValidationService.java:34]: they can't create actually. in case of light duty we create 2 instances, in case of medium we create it with 3 instances. There is a validation in place which would prevent creating freeipa with more than 3 instances

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaTopologyService.java [hunk 5]
```diff
@@ -34,69 +36,75 @@
...
-                    }
-                }
-                String node2 = itr2.next();
-                connectionRemaining.put(node2, connectionRemaining.get(node2) - 1);
-                ret.add(new UnorderedPair(node, node2));
+        Map<String, Integer> connectionRemaining = createConnectionRemainingPerNode(nodes);
+        Set<UnorderedPair> generatedTopology = new HashSet<>();
+        for (String node : new LinkedHashSet<>(connectionRemaining.keySet())) {
+            Integer remainingConnectionForNode = connectionRemaining.get(node);
+            connectionRemaining.remove(node);
```
[the remark this answers is not in the document]
Comment (lacikaaa) [reply freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaTopologyService.java:73]: ok, I'll do that

**Review by horadla23 — APPROVED** [review]
